## Introduction

Online learning has become an essential part of modern education, but student engagement varies widely. Understanding how learners interact with online courses helps educators improve teaching quality, personalize learning, and reduce dropout rates.
In this project, a machine learning model is used to analyze online course engagement data and predict whether a student is highly engaged or not. By applying techniques such as data preprocessing, visualization, and the Random Forest Classifier, the system can identify important patterns in student activity. This helps institutions and instructors make data-driven decisions to improve learning outcomes.

Import Libraries

In [2]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler,LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score,r2_score,mean_squared_error,classification_report,confusion_matrix

Import Dataset

In [3]:
df=pd.read_csv('d:\DataScience\PROJECT\M L project\learn ML\online_course_engagement_data.csv')

Read 1st 5 Rows

In [4]:
df.head()

,UserID,CourseCategory,TimeSpentOnCourse,NumberOfVideosWatched,NumberOfQuizzesTaken,QuizScores,CompletionRate,DeviceType,CourseCompletion
0,5618,Health,29.979719,17,3,50.365656,20.860773,1,0
1,4326,Arts,27.802640,1,5,62.615970,65.632415,1,0
2,5849,Arts,86.820485,14,2,78.458962,63.812007,1,1
3,4992,Science,35.038427,17,10,59.198853,95.433162,0,1
4,3866,Programming,92.490647,16,0,98.428285,18.102478,0,0


Read Last 5 Rows

In [5]:
df.tail()

,UserID,CourseCategory,TimeSpentOnCourse,NumberOfVideosWatched,NumberOfQuizzesTaken,QuizScores,CompletionRate,DeviceType,CourseCompletion
8995,8757,Health,37.445225,14,4,54.469359,32.990704,1,0
8996,894,Science,48.631443,7,7,59.413257,0.254625,0,0
8997,6323,Health,38.212512,3,3,69.508297,70.188159,1,0
8998,3652,Health,70.048665,13,10,79.655182,72.975225,1,1
8999,5595,Health,93.589781,7,5,56.274546,11.299071,0,0


Basic Info

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9000 entries, 0 to 8999
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   UserID                 9000 non-null   int64  
 1   CourseCategory         9000 non-null   object 
 2   TimeSpentOnCourse      9000 non-null   float64
 3   NumberOfVideosWatched  9000 non-null   int64  
 4   NumberOfQuizzesTaken   9000 non-null   int64  
 5   QuizScores             9000 non-null   float64
 6   CompletionRate         9000 non-null   float64
 7   DeviceType             9000 non-null   int64  
 8   CourseCompletion       9000 non-null   int64  
dtypes: float64(3), int64(5), object(1)
memory usage: 632.9+ KB


Describe the Sataset

In [7]:
df.describe()

,UserID,TimeSpentOnCourse,NumberOfVideosWatched,NumberOfQuizzesTaken,QuizScores,CompletionRate,DeviceType,CourseCompletion
count,9000.000000,9000.000000,9000.000000,9000.000000,9000.000000,9000.000000,9000.000000,9000.000000
mean,4498.894556,50.163822,10.024667,5.090556,74.706028,50.340146,0.500667,0.396444
std,2596.849433,28.491750,6.029878,3.157762,14.378383,28.950977,0.500027,0.489186
min,1.000000,1.005230,0.000000,0.000000,50.005119,0.009327,0.000000,0.000000
25%,2251.750000,25.440548,5.000000,2.000000,62.283451,25.653614,0.000000,0.000000
50%,4483.500000,49.818417,10.000000,5.000000,74.743294,50.264124,1.000000,0.000000
75%,6751.250000,75.069924,15.000000,8.000000,87.022663,75.572493,1.000000,1.000000
max,9000.000000,99.992558,20.000000,10.000000,99.994984,99.979711,1.000000,1.000000


Find Null values

In [8]:
df.isna().sum()

UserID                   0
CourseCategory           0
TimeSpentOnCourse        0
NumberOfVideosWatched    0
NumberOfQuizzesTaken     0
QuizScores               0
CompletionRate           0
DeviceType               0
CourseCompletion         0
dtype: int64

Split Data

In [9]:
x=df.drop(columns=['CourseCompletion'])
y=df['CourseCompletion']

Convert Categorical values into numerical

In [10]:
Label_encoder=LabelEncoder()
for col in x.select_dtypes(include=['object']).columns:
    x[col]=Label_encoder.fit_transform(x[col])

Split into Train And Test Sets

In [11]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

Train model usiing Random Forest Classifier

In [12]:
model=RandomForestClassifier(random_state=42)
model.fit(x_train,y_train)

RandomForestClassifier(random_state=42)

Evaluate the Accuracy

In [ ]:
y_pred=model.predict(x_test)
accuracy=accuracy_score(y_test,y_pred)
print(f"Random Forest Model Accuracy:{accuracy*100:.2f}%\n")

Random Forest Model Accuracy:96.22%



## Conclusion

This project successfully demonstrates how machine learning can be used to predict student engagement in online courses. After training and testing the model, the Random Forest Classifier achieved a strong accuracy of about 96%, showing its effectiveness in identifying engagement levels.
The results highlight that data-driven approaches can help educational platforms understand learner behavior, provide early intervention, and enhance the overall learning experience. With further improvements and more data, this system can become a powerful tool for improving online education quality.